File: src/limitedwip/watchdog/Watchdog.java
package limitedwip.watchdog;

import limitedwip.watchdog.components.IdeAdapter;

public class Watchdog {
	private static final int undefined = -1;

	private final IdeAdapter ideAdapter;
    private Settings settings;
    private int lastNotificationTime = undefined;
    private boolean skipNotificationsUtilCommit = false;


    public Watchdog(IdeAdapter ideAdapter) {
        this.ideAdapter = ideAdapter;
    }

	public Watchdog init(Settings settings) {
		onSettings(settings);
		return this;
	}

	public void onTimer(int seconds) {
        if (!settings.enabled) return;

        ChangeSize changeSize = ideAdapter.currentChangeListSizeInLines();
		boolean exceededThreshold = changeSize.value > settings.maxLinesInChange;
		boolean timeToNotify =
		        lastNotificationTime == undefined ||
		        (seconds - lastNotificationTime) >= settings.notificationIntervalInSeconds;

		if (timeToNotify && exceededThreshold && !skipNotificationsUtilCommit) {
			ideAdapter.onChangeSizeTooBig(changeSize, settings.maxLinesInChange);
			lastNotificationTime = seconds;
		}
		if (!exceededThreshold){
			ideAdapter.onChangeSizeWithinLimit();
		}

		ideAdapter.showCurrentChangeListSize(changeSize, settings.maxLinesInChange);
	}

    public void onSettings(Settings settings) {
	    ideAdapter.onSettingsUpdate(settings);
	    lastNotificationTime = undefined;
        this.settings = settings;
    }

    public void onCommit() {
	    // This is a workaround to suppress notifications sent while commit dialog is open.
	    ideAdapter.onChangeSizeWithinLimit();

	    skipNotificationsUtilCommit = false;
    }

    public void skipNotificationsUntilCommit(boolean value) {
        skipNotificationsUtilCommit = value;
	    ideAdapter.onSkipNotificationUntilCommit(value);
    }

    public void toggleSkipNotificationsUntilCommit() {
        skipNotificationsUntilCommit(!skipNotificationsUtilCommit);
    }


    public static class Settings {
	    public final boolean enabled;
	    public final int maxLinesInChange;
	    public final int notificationIntervalInSeconds;
	    public final boolean showRemainingChangesInToolbar;

        public Settings(boolean enabled, int maxLinesInChange, int notificationIntervalInSeconds, boolean showRemainingChangesInToolbar) {
            this.enabled = enabled;
            this.maxLinesInChange = maxLinesInChange;
            this.notificationIntervalInSeconds = notificationIntervalInSeconds;
	        this.showRemainingChangesInToolbar = showRemainingChangesInToolbar;
        }
    }
}

File: src/limitedwip/watchdog/Watchdog.kt
package limitedwip.watchdog

import limitedwip.common.PathMatcher
import limitedwip.common.settings.LimitedWipSettings.Companion.never
import limitedwip.watchdog.components.WatchdogIde

class Watchdog(private val ide: WatchdogIde, private var settings: Settings) {
    private var secondsTillNotification = 0
    private var skipNotificationsUntilCommit = false
    private var allowOneCommitWithoutChecks = false

    init {
        onSettingsUpdate(settings)
    }

    fun onTimer() {
        if (!settings.enabled) return

        ide.calculateCurrentChangeListSizeInLines()

        val changeSize = ide.currentChangeListSizeInLines().applyExclusions()
        val exceededThreshold = changeSize.value > settings.maxLinesInChange

        secondsTillNotification--
        if (exceededThreshold && settings.notificationIntervalInSeconds != never && secondsTillNotification <= 0 && !skipNotificationsUntilCommit) {
            ide.showNotificationThatChangeSizeIsTooBig(changeSize, settings.maxLinesInChange)
            secondsTillNotification = settings.notificationIntervalInSeconds
        }
        if (!exceededThreshold) {
            ide.hideNotificationThatChangeSizeIsTooBig()
        }

        ide.showCurrentChangeListSize(changeSize, settings.maxLinesInChange)
    }

    fun onSettingsUpdate(settings: Settings) {
        ide.onSettingsUpdate(settings)
        secondsTillNotification = 0 // Set to 0 so that notifications are sent immediately after watchdog is started.
        if (this.settings.maxLinesInChange != settings.maxLinesInChange) {
            ide.showCurrentChangeListSize(ide.currentChangeListSizeInLines().applyExclusions(), settings.maxLinesInChange)
        }
        this.settings = settings
    }

    fun onSuccessfulCommit() {
        // This is a workaround to suppress notifications sent while commit dialog is open.
        ide.hideNotificationThatChangeSizeIsTooBig()

        skipNotificationsUntilCommit = false
        allowOneCommitWithoutChecks = false
    }

    fun isCommitAllowed(changeSizesWithPath: ChangeSizesWithPath): Boolean {
        if (allowOneCommitWithoutChecks || !settings.noCommitsAboveThreshold || !settings.enabled) return true
        if (changeSizesWithPath.applyExclusions().value > settings.maxLinesInChange) {
            ide.notifyThatCommitWasCancelled()
            return false
        }
        return true
    }

    fun toggleSkipNotificationsUntilCommit() {
        skipNotificationsUntilCommit(!skipNotificationsUntilCommit)
    }

    fun onSkipNotificationsUntilCommit() {
        skipNotificationsUntilCommit(true)
    }

    fun onForceCommit() {
        allowOneCommitWithoutChecks = true
        ide.commitWithoutDialog()
    }

    private fun ChangeSizesWithPath.applyExclusions(): ChangeSize {
        return ChangeSizesWithPath(
            value.filter { (path, _) -> settings.exclusions.none { it.matches(path) } }
        ).totalChangeSize
    }

    private fun skipNotificationsUntilCommit(value: Boolean) {
        skipNotificationsUntilCommit = value
        ide.showNotificationThatWatchdogIsDisableUntilNextCommit(value)
    }

    data class Settings(
        val enabled: Boolean,
        val maxLinesInChange: Int,
        val notificationIntervalInSeconds: Int,
        val showRemainingChangesInToolbar: Boolean,
        val noCommitsAboveThreshold: Boolean,
        val exclusions: Set<PathMatcher>
    )
}
